### PR TITLE
WIP move ordered iterators to associated items

### PR DESCRIPTION
**Pull request by Gankro**
Doesn't work because of this ICE: https://github.com/rust-lang/rust/issues/20220

But I think this is the right approach.

### CODE REVIEW

**Comment by csherratt** [comment]
:o this is much nicer.

**Comment by Gankro** [comment]
Bah, the ICE persists

File: src/iter/ordered_iter.rs [hunk 1]
```diff
@@ -243,29 +252,43 @@ impl<K: Ord+Eq,
...
+impl<'a, K: Ord, V> OrderedMapIterator<Key=&'a K, Val=&'a V> for btree_map::Iter<'a, K, V> {}
 #[cfg(feature="tree_map")]
-impl<'a, K> OrderedSetIterator<&'a K> for tree_set::Iter<'a, K> {}
+impl<'a, K> OrderedSetIterator<Item=&'a K> for tree_set::Iter<'a, K> {}
 #[cfg(feature="tree_map")]
-impl<'a, K, V> OrderedMapIterator<&'a K, &'a V> for tree_map::Iter<'a, K, V> {}
-#[cfg(feature="trie_map")]
-impl<'a> OrderedSetIterator<usize> for trie_set::Iter<'a> {}
+impl<'a, K, V> OrderedMapIterator<Key=&'a K, Val=&'a V> for tree_map::Iter<'a, K, V> {}
+#[cfg(feature="tree_map")]
```
Comment (apasel422) [inline src/iter/ordered_iter.rs]: s/tree_map/trie_map/

File: src/iter/ordered_iter.rs [hunk 2]
```diff
@@ -243,29 +252,43 @@ impl<K: Ord+Eq,
     }
 }
 
-impl<'a, K: Ord> OrderedSetIterator<&'a K> for btree_set::Iter<'a, K> {}
-impl<'a, K: Ord, V> OrderedMapIterator<&'a K, &'a V> for btree_map::Iter<'a, K, V> {}
+impl<'a, K: Ord> OrderedSetIterator<Key=&'a K> for btree_set::Iter<'a, K> {}
```
Comment (apasel422) [inline src/iter/ordered_iter.rs]: These lines use the wrong syntax for associated types, and anyway, `OrderedSetIterator` no longer has an associated type (it uses `Iterator::Item`), per the changes above. Lines 255-256, for example, should be changed to
```rust
impl<'a, K: Ord> OrderedSetIterator for btree_set::Iter<'a, K> {}

impl<'a, K: Ord, V> OrderedMapIterator for btree_map::Iter<'a, K, V> {
    type Key = &'a K;
    type Val = &'a V;
}
```

File: src/iter/ordered_iter.rs [hunk 3]
```diff
@@ -243,29 +252,43 @@ impl<K: Ord+Eq,
...
-     B: OrderedMapIterator<K, VB>> OrderedMapIterator<K, (VA, VB)> for InnerJoinMapIterator<A, B> {}
-impl<K: Ord, V,
-     A: OrderedSetIterator<K>,
-     B: OrderedMapIterator<K, V>> OrderedMapIterator<K, V> for InnerJoinMapSetIterator<B, A> {}
-impl<K: Ord,
-     A: OrderedSetIterator<K>,
-     B: OrderedSetIterator<K>> OrderedSetIterator<K> for InnerJoinSetIterator<A, B> {}
+impl<'a, V> OrderedMapIterator<Item=usize, Val=&'a V> for trie_map::Iter<'a, V> {}
+impl<'a, V> OrderedMapIterator<Key=usize, Val=&'a V> for vec_map::Iter<'a, V> {}
+impl<'a> OrderedSetIterator<Item=usize> for bitv_set::Iter<'a> {}
```
Comment (apasel422) [inline src/iter/ordered_iter.rs]: s/bitv_set/bit_set/

File: src/iter/ordered_iter.rs (hunk 2 above)
Comment (Gankro) [inline src/iter/ordered_iter.rs]: Oh duh. The ICE never let me get this far :(

**Comment by Gankro** [comment]
Current roadblock:

```

**Comment by apasel422** [comment]
Oh, yeah. That is https://github.com/rust-lang/rust/issues/20825.

**Comment by apasel422** [comment]
If you rebase on master, this will build now. You can add
```toml
[profile.dev]
debug = false
```
to `Cargo.toml` to avoid the issue I referenced earlier. While you're at it, we can remove `#![feature(old_impl_check)]` from `lib.rs`.

**Comment by Gankro** [comment]
I'm still seeing the error. Do you mean rebase on rust master? As in the next nightly?

**Comment by apasel422** [comment]
Using today's nightly (`rustc 1.0.0-nightly (b0746ff19 2015-03-05) (built 2015-03-06)`), I rebased your commits on `collect-rs/master` and it built and tested successfully after I modified `Cargo.toml` as above.

**Comment by Gankro** [comment]
Ah I'm a nightly back. On a bad internet connection, want to finish this PR up?

**Comment by apasel422** [comment]
Sure. Let me take a look.

On Fri, Mar 6, 2015 at 7:16 PM, Alexis Beingessner <notifications@github.com

**Comment by Gankro** [comment]
:confetti_ball:

**Pull request merged**